README.md
# 🌲 Fluent::Plugin::Timber, a plugin for [Fluentd](http://fluentd.org)

[![Gem Version](https://badge.fury.io/rb/fluent-plugin-timber.png)](http://badge.fury.io/rb/fluent-plugin-timber)
[![Build Status](https://travis-ci.org/timberio/fluent-plugin-timber.png?branch=master)](https://travis-ci.org/timberio/fluent-plugin-timber)
[![Issue Stats](http://issuestats.com/github/uken/fluent-plugin-timber/badge/pr)](http://issuestats.com/github/uken/fluent-plugin-timber)
[![Issue Stats](http://issuestats.com/github/uken/fluent-plugin-timber/badge/issue)](http://issuestats.com/github/uken/fluent-plugin-timber)

A highly efficient Fluentd plugin that delivers events to the [Timber.io logging service](https://timber.io). It uses batching, msgpack, and retry logic for highly efficient and reliable delivery of log data.

[Timber.io](https://timber.io) is a different kind of logging service with a focus on modern logging best-practices: easy setup, structured data, fast clean usable interface, 6 months of searchable retention, threshold based alerts, simple graphing, and more. Learn more at [https://timber.io](https://timber.io).

## Installation

```
gem install fluent-plugin-timber
```

## Usage

In your Fluentd configuration, use @type timber:

```
<match your_match>
  @type timber
  api_key xxxxxxxxxxxxxxxxxxxxxxxxxxx        # Your Timber API (required)
  hostname "#{Socket.gethostname}"           # Your hostname (required)
  # ip 127.0.0.1                             # IP address (optional)
  buffer_chunk_limit 1m                      # Must be < 5m
  flush_at_shutdown true                     # Only needed with file buffer
</match>
```

## Configuration

* `api_key` - This is your Timber API key. You can obtain your key by creating an app in the [Timber console](https://app.timber.io). Registration is one click. If you already have an app, you can locate your API in your app's settings. [See our API key docs](https://timber.io/docs/app/advanced/api-keys/).
* `hostname` - This adds `hostname` as context to your logs, making it easy to filter by hostname in the [Timber console](https://app.timber.io).
* `ip` - This adds `ip` as context to your logs, making it easy to filter by IP address in the [Timber console](https://app.timber.io).

For advanced configuration options, please see to the [buffered output parameters documentation.](http://docs.fluentd.org/articles/output-plugin-overview#buffered-output-parameters).

---

Questions? Need help? [support@timber.io](mailto:support@timber.io).

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## Unreleased

## 1.0.2 - 2019-02-23

### Changed

  - Updated the logs endpoint from `/sources/:source_id` to `/sources/:source_id/frames`

spec/fluent/plugin/out_timber_spec.rb
require "spec_helper"
require "fluent/plugin/out_timber"

describe Fluent::TimberOutput do
  let(:config) do
    %{
      api_key  abcd1234
      hostname my.host.com
    }
  end

  let(:driver) do
    tag = "test"
    Fluent::Test::BufferedOutputTestDriver.new(Fluent::TimberOutput, tag) {
      # v0.12's test driver assume format definition. This simulates ObjectBufferedOutput format
      if !defined?(Fluent::Plugin::Output)
        def format(tag, time, record)
          [time, record].to_msgpack
        end
      end
    }.configure(config)
  end
  let(:record) do
    {'age' => 26, 'request_id' => '42', 'parent_id' => 'parent', 'routing_id' => 'routing'}
  end

  before(:each) do
    Fluent::Test.setup
  end

  describe "#write" do
    it "should send a chunked request to the Timber API" do
      stub = stub_request(:post, "https://logs.timber.io/frames").
        with(
          :body => start_with("\x85\xA3age\x1A\xAArequest_id\xA242\xA9parent_id\xA6parent\xAArouting_id\xA7routing\xA2dt\xB4".force_encoding("ASCII-8BIT")),
          :headers => {'Authorization'=>'Basic YWJjZDEyMzQ=', 'Connection'=>'Keep-Alive', 'Content-Type'=>'application/msgpack', 'User-Agent'=>'Timber Logstash/1.0.0'}
        ).
        to_return(:status => 200, :body => "", :headers => {})

      driver.emit(record)
      driver.run

      expect(stub).to have_been_requested.times(1)
    end

    it "handles 500s" do
      stub = stub_request(:post, "https://logs.timber.io/frames").to_return(:status => 500, :body => "", :headers => {})

      driver.emit(record)
      driver.run

      expect(stub).to have_been_requested.times(3)
    end

    it "handle auth failures" do
      stub = stub_request(:post, "https://logs.timber.io/frames").to_return(:status => 403, :body => "", :headers => {})

      driver.emit(record)
      driver.run

      expect(stub).to have_been_requested.times(1)
    end
  end
end

lib/fluent/plugin/out_timber.rb
require 'base64'
require 'json'

require 'fluent/output'

module Fluent
  class TimberOutput < Fluent::BufferedOutput
    Fluent::Plugin.register_output('timber', self)

    VERSION = "1.0.1".freeze
    CONTENT_TYPE = "application/msgpack".freeze
    HOST = "https://logs.timber.io".freeze
    MAX_ATTEMPTS = 3.freeze
    PATH = "/frames".freeze
    RETRYABLE_CODES = [429, 500, 502, 503, 504].freeze
    USER_AGENT = "Timber Logstash/#{VERSION}".freeze

    config_param :api_key, :string, secret: true
    config_param :hostname, :string
    config_param :ip, :string, default: nil

    def configure(conf)
      source_id = conf["source_id"]
      api_key = conf["api_key"]
      @path = "/sources/#{source_id}/frames"
      @headers = {
        "Authorization" => "Bearer #{api_key}",
        "Content-Type" => CONTENT_TYPE,
        "User-Agent" => USER_AGENT
      }
      super
    end

    def start
      super
      require 'http'
      HTTP.default_options = {:keep_alive_timeout => 29}
      @http_client = HTTP.persistent(HOST)
    end

    def shutdown
      @http_client.close if @http_client
      super
    end

    def format(tag, time, record)
      dt_iso8601 = Time.at(time).utc.iso8601
      record.merge("dt" => dt_iso8601).to_msgpack
    end

    def write(chunk)
      deliver(chunk, 1)
    end

    private
      def deliver(chunk, attempt)
        if attempt > MAX_ATTEMPTS
          log.error("msg=\"Max attempts exceeded dropping chunk\" attempt=#{attempt}")
          return false
        end

        body = chunk.read
        response = @http_client.headers(@headers).post(@path, body: body)
        response.flush
        code = response.code

        if code >= 200 && code <= 299
          true
        elsif RETRYABLE_CODES.include?(code)
          sleep_time = sleep_for_attempt(attempt)
          log.warn("msg=\"Retryable response from the Timber API\" " +
            "code=#{code} attempt=#{attempt} sleep=#{sleep_time}")
          sleep(sleep_time)
          deliver(chunk, attempt + 1)
        else
          log.error("msg=\"Fatal response from the Timber API\" code=#{code} attempt=#{attempt}")
          false
        end
      end

      def sleep_for_attempt(attempt)
        sleep_for = attempt ** 2
        sleep_for = sleep_for <= 60 ? sleep_for : 60
        (sleep_for / 2) + (rand(0..sleep_for) / 2)
      end
  end
end